### PR TITLE
Added masking type index_t for all references to hnswlib indices (als…

### PR DESCRIPTION
**Pull request by bwpriest**
…o called labels). Replaced prior interchangeable use of size_t and uint64_t.

### CODE REVIEW

**Comment by bwpriest** [comment]
I have bundled all references (I think I got them all) to the point indices under the defined type `saltatlas::index_t`. The original code was using `std::size_t` and `std::uint64_t` interchangeably, which is not portable. Right now I am setting `saltatlas::index_t = std::size_t` so as to leave the semantics unchanged. We could instead add a `IndexType` template parameter to all of the classes/functions that currently use `saltatlas::index_t`, but I am unsure how useful it would be to support different index types. What are your thoughts, @steiltre?

**Comment by steiltre** [comment]
Thanks Ben. Templating on an `IndexType` with a default would be consistent with the [dnnd implementation](https://github.com/LLNL/saltatlas/blob/develop/include/saltatlas/dnnd/dnnd.hpp#L31).

I can't say with certainty, but some of the data we want to use with this may have a use for different `IndexType`. The requirements shouldn't be too restrictive either based on how they're used in the code. Being able to use them as keys in a map should be sufficient (might need to be hashable as well).

**Comment by bwpriest** [comment]
Ok. I'll update this PR with a commit moving to a templating strategy this afternoon.

**Comment by bwpriest** [comment]
Sorry, that took a little longer than expected. I've made the `IndexType` a template parameter throughout. I also made `Partitioner` a template class parameter, so you can now do things like

```
metric_hyperplane_partitioner<float, std::size_t, std::vector<float>> partitioner(world, my_l2_space);
dhnsw dist_index(voronoi_rank, num_seeds, &my_l2_space, &world, partitioner);
```

The template arguments to `metric_hyperplane_partitioner` get reused by dhnsw, so it is not necessary to apply them twice. This also makes it harder to accidentally mess up and use incompatible types since you only specify them once.

I also cleaned up the interiors of the classes with defined types. I definitely made some strong choices in the last commit, so let me know if I went too far @steiltre.

**Comment by bwpriest** [comment]
I decided to go ahead and use a template class for the partitioning interface, which now expects a `template<template <typename IndexType, typename Point> class Container>` object. This allows us to verify that the partitioner will iterate over `IndexType`-`Point` value pairs. However, to get the examples to compile I had to make a `pair_bag` class which has two template arguments that it passes along to `std::pair`values. It is otherwise a verbatim copy of `ygm::container::bag`. I am certain that there is more elegant way to do this, but should suffice for now.

I wanted to use
```
template <typename IndexType, typename Point>
using pair_bag = ygm::container::bag<std::pair<IndexType, Point>>;
```  
but the compiler does not see this as a distinct type. It untagles the templates such that when `saltatlas::dhnsw::partition_data()` accepts an argument of type `pair_bag`, it sees an argument of type `ygm::container::bag`.

File: include/saltatlas/dhnsw/dhnsw.hpp [hunk 1]
```diff
@@ -72,9 +79,9 @@ m_index_impl.store_seeds(seed_features);
   inline ygm::comm &comm() { return *m_comm; }
 
  private:
-  ygm::comm                                             *m_comm;
-  dhnsw_detail::dhnsw_impl<DistType, Point, Partitioner> m_index_impl;
-  dhnsw_detail::query_engine_impl<DistType, Point, Partitioner>
+  ygm::comm                                                      *m_comm;
+  dhnsw_detail::dhnsw_impl<dist_t, index_t, point_t, Partitioner> m_index_impl;
+  dhnsw_detail::query_engine_impl<dist_t, index_t, point_t, Partitioner>
```
Comment (steiltre) [inline include/saltatlas/dhnsw/dhnsw.hpp:84]: Would be nice if this could be partitioner_t instead of Partitioner.

**Comment by steiltre** [comment]
> I decided to go ahead and use a template class for the partitioning interface, which now expects a `template<template <typename IndexType, typename Point> class Container>` object. This allows us to verify that the partitioner will iterate over `IndexType`-`Point` value pairs. However, to get the examples to compile I had to make a `pair_bag` class which has two template arguments that it passes along to `std::pair`values. It is otherwise a verbatim copy of `ygm::container::bag`. I am certain that there is more elegant way to do this, but should suffice for now.
> 
> I wanted to use
> 
> ```
> template <typename IndexType, typename Point>
> using pair_bag = ygm::container::bag<std::pair<IndexType, Point>>;
> ```
> 
> but the compiler does not see this as a distinct type. It untagles the templates such that when `saltatlas::dhnsw::partition_data()` accepts an argument of type `pair_bag`, it sees an argument of type `ygm::container::bag`.

Not having to make a `pair_bag` type would be nice. Having it built to use the `ygm::container::detail::bag_impl` isn't the worst thing ever. I think this would work replacing `template <typename, typename> class Container` with just `class Container` would work because the `for_all` for a `ygm::container::map` and `ygm::container::bag<std::pair<>>` buried within the partitioners would both be iterating over pairs. This would definitely be more fragile and require more memorization of what's happening though.

**Review by steiltre — APPROVED** [review]

**Comment by steiltre** [comment]
Thanks for the changes. Going to merge here and merge into develop on the main repo.

**Comment by bwpriest** [comment]
> > I decided to go ahead and use a template class for the partitioning interface, which now expects a `template<template <typename IndexType, typename Point> class Container>` object. This allows us to verify that the partitioner will iterate over `IndexType`-`Point` value pairs. However, to get the examples to compile I had to make a `pair_bag` class which has two template arguments that it passes along to `std::pair`values. It is otherwise a verbatim copy of `ygm::container::bag`. I am certain that there is more elegant way to do this, but should suffice for now.
> > I wanted to use
> > ```
> > template <typename IndexType, typename Point>
> > using pair_bag = ygm::container::bag<std::pair<IndexType, Point>>;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but the compiler does not see this as a distinct type. It untagles the templates such that when `saltatlas::dhnsw::partition_data()` accepts an argument of type `pair_bag`, it sees an argument of type `ygm::container::bag`.
> 
> Not having to make a `pair_bag` type would be nice. Having it built to use the `ygm::container::detail::bag_impl` isn't the worst thing ever. I think this would work replacing `template <typename, typename> class Container` with just `class Container` would work because the `for_all` for a `ygm::container::map` and `ygm::container::bag<std::pair<>>` buried within the partitioners would both be iterating over pairs. This would definitely be more fragile and require more memorization of what's happening though.

I concur that the `pair_bag` type is overengineered. I'm a little leery of using a raw template type though, because ideally I'd like the compiler to know that the `for_all` invocation needs to loop over key-value pairs. I feel like there must be a way to declare a type alias to the compiler as I described above, but I have not yet figured out how to do so.

**Comment by bwpriest** [comment]
It looks like what I was trying to do is use a strong typedef, for which there are many custom solutions (including a simple one for built in types in Boost) but no built-in mechanism. I'll leave the `pair_bag` class as it is for now, but if we find ourselves needing more of this functionality in the future I will revisit. For example, it might be convenient for the compiler to know not to convert an `IndexType` object to a compatible integer type and vice versa.

**Pull request merged**